### PR TITLE
Add and fix compile warnings

### PR DESCRIPTION
**Pull request by gitbuda**
NOTE: Warnings as errors are disabled by default because compilation on the latest compilers might fail because of a newly added warning (or something similar) which might confuse users. On the other hand, within CI, all warnings should be errors.

- [x] Add warnings per compiler per language.
- [x] Fix as many errors as possible.
- [x] Make AppleClang warnings errors.

### CODE REVIEW

File: cmake/CompilerWarnings.cmake [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      -Wpedantic
+  )
+
+  if (C_WARNINGS_AS_ERRORS)
+    set(CLANG_C_WARNINGS ${CLANG_C_WARNINGS} -Werror)
+    set(GCC_C_WARNINGS ${GCC_C_WARNINGS} -Werror)
+  endif()
+
+  if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    set(C_PROJECT_WARNINGS ${CLANG_C_WARNINGS})
```
Comment (gitbuda) [inline cmake/CompilerWarnings.cmake:21]: Add `AppleClang` here as well.

File: tests/bolt-testdata.hpp [hunk 2]
```diff
@@ -261,9 +264,9 @@ std::vector<ValueTestParam> NodeTestCases() {
   std::vector<ValueTestParam> inputs;
 
   {
-    mg_string *labels[0] = {};
+    std::vector<mg_string *> labels(0);
```
Comment (antonio2368) [inline tests/bolt-testdata.hpp:267]: What kind of error did you get here?

File: tests/decoder.cpp [hunk 3]
```diff
@@ -278,57 +278,64 @@ TEST_P(ValueTest, Decoding) {
...
+INSTANTIATE_TEST_CASE_P(Null, ValueTest,
+                        ::testing::ValuesIn(NullTestCases()), );
 
-INSTANTIATE_TEST_CASE_P(Bool, ValueTest, ::testing::ValuesIn(BoolTestCases()));
+INSTANTIATE_TEST_CASE_P(Bool, ValueTest,
+                        ::testing::ValuesIn(BoolTestCases()), );
 
 INSTANTIATE_TEST_CASE_P(Integer, ValueTest,
-                        ::testing::ValuesIn(IntegerTestCases()));
+                        ::testing::ValuesIn(IntegerTestCases()), );
```
Comment (antonio2368) [inline tests/decoder.cpp:288]: Why `, )` ?

File: tests/encoder.cpp [hunk 4]
```diff
@@ -100,13 +100,13 @@ class EncoderTest : public ::testing::Test {
 };
 
 void AssertReadRaw(std::stringstream &sstr, const std::string &expected) {
-  char tmp[expected.size()];
-  size_t read = sstr.readsome(tmp, expected.size());
+  std::vector<char> tmp(expected.size());
```
Comment (antonio2368) [inline tests/encoder.cpp:103]: Same question as before, what was the warning here?

File: tests/unit/mgclient_value.cpp [hunk 5]
```diff
@@ -230,11 +230,11 @@ TEST(ValueTest, MapIterate) {
   map.Insert("key 3", Value(3.0));
 
   std::vector<std::pair<std::string, ConstValue>> values;
-  for (const auto &[key, value] : map) {
+  for (const auto [key, value] : map) {
```
Comment (antonio2368) [inline tests/unit/mgclient_value.cpp:233]: Why did you remove reference?
You can move now key and value in the emplace_back call bellow but don't see any other reason why you would do this now?

File: tests/bolt-testdata.hpp (hunk 2 above)
Comment (antonio2368) [reply tests/bolt-testdata.hpp:267]: Oh, C++ doesn't like VLA, good to know.

Comment (gitbuda) [reply tests/bolt-testdata.hpp:267]: Yep, people even say VLAs are a huge mistake in the C standard...

File: tests/unit/mgclient_value.cpp (hunk 5 above)
Comment (gitbuda) [reply tests/unit/mgclient_value.cpp:233]: AppleClang 12 says
```
[ 62%] Building CXX object tests/CMakeFiles/unit_mgclient_value.dir/unit/mgclient_value.cpp.o
/Users/runner/work/mgclient/mgclient/tests/unit/mgclient_value.cpp:121:20: warning: loop variable 'value' is always a copy because the range of type 'mg::List' does not return a reference [-Wrange-loop-analysis]
  for (const auto &value : list) {
                   ^
/Users/runner/work/mgclient/mgclient/tests/unit/mgclient_value.cpp:121:8: note: use non-reference type 'mg::ConstValue'
  for (const auto &value : list) {
       ^~~~~~~~~~~~~~~~~~~
```

File: tests/encoder.cpp (hunk 4 above)
Comment (gitbuda) [reply tests/encoder.cpp:103]: Yea, the same as above, any alternative to the vector? But I think it's pretty unimportant in general...

File: tests/decoder.cpp (hunk 3 above)
Comment (gitbuda) [reply tests/decoder.cpp:288]: Clang says
```
/home/buda/Workspace/code/memgraph/mgclient/tests/decoder.cpp:282:61: error: must specify at least one argument for '...' parameter of variadic macro [-Werror,-Wgnu-zero-variadic-macro-arguments]
                        ::testing::ValuesIn(NullTestCases()));
                                                            ^
/home/buda/Workspace/code/memgraph/mgclient/build/gtest/include/gtest/gtest-param-test.h:1413:10: note: macro 'INSTANTIATE_TEST_CASE_P' defined here
# define INSTANTIATE_TEST_CASE_P(prefix, test_case_name, generator, ...) 
```

File: tests/bolt-testdata.hpp (hunk 2 above)
Comment (gitbuda) [reply tests/bolt-testdata.hpp:267]: Actually, the problem here is a 0 sized array...

File: tests/encoder.cpp (hunk 4 above)
Comment (antonio2368) [reply tests/encoder.cpp:103]: Yeah, it's not important just was curious. Vector should be fine or:
`auto tmp = std::make_unique<char[]>(expected.size())`?

File: tests/unit/mgclient_value.cpp (hunk 5 above)
Comment (antonio2368) [reply tests/unit/mgclient_value.cpp:233]: That's a nice warning.

Comment (gitbuda) [reply tests/unit/mgclient_value.cpp:233]: Yep :smile:

File: tests/encoder.cpp (hunk 4 above)
Comment (gitbuda) [reply tests/encoder.cpp:103]: Let's leave vector then :smile:

**Review by antonio2368 — APPROVED** [review]
